frontend/public/components/modals/taint-update-modal.tsx

```diff
@@ -0,0 +1,140 @@
+    ).then(this.props.close, () => {});
+  };
+
+  _cancel = () => {
+    this.props.close();
+  };
+
+  _change = (e, i, field) => {
+    const newValue = e.target ? e.target.value : e;
+    this.setState((prevState: TaintUpdateModalState) => {
+      const clonedTaints = _.cloneDeep(prevState.taints);
+      clonedTaints[i][field] = newValue;
+      return {
+        taints: clonedTaints,
+      };
+    });
+  };
+
+  _remove = (i) => {
+    this.setState((prevState: TaintUpdateModalState) => {
+      const clonedTaints = _.cloneDeep(this.state.taints);
+      clonedTaints.splice(i, 1);
+      return {
+        taints: clonedTaints,
+      };
+    });
+  };
+
+  _addRow = () => {
+    this.setState((prevState: TaintUpdateModalState) => ({
+      taints: [...prevState.taints, {key: '', value: '', effect: ''}],
+    }));
+  };
+
+  render() {
+    const effects = {
+      '': '',
+      'NoSchedule': 'NoSchedule',
+      'PreferNoSchedule': 'PreferNoSchedule',
+      'NoExecute': 'NoExecute',
+    };
+    return <form onSubmit={this._submit} name="form" className="modal-content taint-update-modal">
+      <ModalTitle>Edit Taints</ModalTitle>
+      <ModalBody>
+        <div className="row hidden-sm hidden-xs">
+          <div className="col-sm-4 text-secondary text-uppercase">Key</div>
+          <div className="col-sm-3 text-secondary text-uppercase">Value</div>
+          <div className="col-sm-4 text-secondary text-uppercase">Effect</div>
+          <div className="col-sm-1 co-empty__header"></div>
+        </div>
+        {_.map(this.state.taints, (c, i) =>
+          <div className="row">
+            <div className="col-md-4 col-sm-6 col-xs-6 taint-update-modal__field">
+              <div className="hidden-md hidden-lg text-secondary text-uppercase">Key</div>
+              <input type="text" className="form-control" value={c.key} onChange={(e) => this._change(e, i, 'key')} />
+            </div>
+            <div className="col-md-3 col-sm-5 col-xs-5 taint-update-modal__field">
+              <div className="hidden-md hidden-lg text-secondary text-uppercase">Value</div>
+              <input type="text" className="form-control" value={c.value} onChange={(e) => this._change(e, i, 'value')} />
+            </div>
+            <div className="clearfix visible-sm visible-xs"></div>
+            <div className="col-md-4 col-sm-6 col-xs-6 taint-update-modal__field">
+              <div className="hidden-md hidden-lg text-secondary text-uppercase">Effect</div>
+              <Dropdown
+                className="taint-update-modal__dropdown"
+                dropDownClassName="dropdown--full-width"
+                id={`taint_effect_dropdown_${i}`}
+                items={effects}
+                onChange={(e) => this._change(e, i, 'effect')}
+                selectedKey={c.effect} />
+            </div>
+            <div className="hidden-lg hidden-md col-sm-5 col-xs-5"></div>
```

Comment:
It's unclear to me why we have the empty divs

Reply:
They are helping with responsive formatting. Since the form uses bootstrap grid for formatting, the padding put on the fields makes it so that nesting columns doesn't lay everything out well, so I used blank divs as spacers so everything lines up when there is an odd number of fields on the form.

Reply:
OK, should we be using column offsets instead?

https://getbootstrap.com/docs/3.3/css/#grid-offsetting

Reply:
ah I'll look at that closer